core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -161,7 +212,23 @@ private JoinNode getSizeBasedJoin(JoinNode joinNode, Context context)
 
         if (isLeftSideSmall) {
             // left side is small enough, but must be partitioned
-            return joinNode.flipChildren().withDistributionType(PARTITIONED);
+            return flippedJoin.withDistributionType(PARTITIONED);
+        }
+
+        // Flip join sides if one side is smaller than the other by more than 8 times.
+        // We use the 8x factor because the reduction in size from aggregation and filtering
+        // may not be fully accounted here.
+        if (rightSourceSize * 8 < leftSourceSize) {
+            if (mustReplicate(joinNode, context)) {
+                return joinNode.withDistributionType(REPLICATED);
+            }
+            return joinNode.withDistributionType(PARTITIONED);
+        }
+        else if (leftSourceSize * 8 < rightSourceSize) {
```

Comment:
Question: Should we use `getBestKnownRowCount` (less restrictive) for flipping join sides and previous implementation of `getSourceTablesSizeInBytes` (more restrictive) for broadcast join selection?

Rationale: choosing broadcast join when we shouldn't will cause bigger perf impact and since we already couldn't estimate input stats, `getBestKnownRowCount` stats are potentially not estimated correctly (underestimated, e.g. due to correlation or default filter factor).

`getBestKnownRowCount` might be better for join side selection because in case of error we will increase memory usage and latency, but it's not as bad as broadcast join which shouldn't happen

Reply:
Yes, I was thinking of doing that as well, I'll switch to using the `getBestKnownRowCount` only for the partitioned join side selection case.